metaflow/plugins/kfp/s3_sensor_decorator.py

```diff
@@ -0,0 +1,109 @@
+from metaflow.decorators import FlowDecorator
+from metaflow.exception import MetaflowException
+
+from types import FunctionType
+from typing import Tuple
+
+"""
+Within identity_formatter, which is passed in as the func parameter,
+customers have access to all variables in flow_parameters_json (which 
+include all parameters passed in through the flow at compile and run time by
+KFP) as well as a number of environment variables, including:
+
+MF_POD_NAME
+MF_POD_NAMESPACE
+MF_ARGO_NODE_NAME
+MF_ARGO_WORKFLOW_NAME
+ZODIAC_SERVICE
+ZODIAC_TEAM
+POD_NAMESPACE
+KFP_SDK_NAMESPACE
+POD_NAME
+POD_IP
+ARGO_WORKFLOW_NAME
+K8S_CLUSTER_NAME
+K8S_CLUSTER_ENV
+METAFLOW_NOTIFY_EMAIL_SMTP_HOST
+METAFLOW_NOTIFY_EMAIL_SMTP_PORT
+METAFLOW_NOTIFY_EMAIL_FROM
+METAFLOW_NOTIFY_EMAIL_BODY
+METAFLOW_NOTIFY_ON_ERROR
+KFP_RUN_URL_PREFIX
+METAFLOW_DEFAULT_DATASTORE
+METAFLOW_DEFAULT_METADATA
+METAFLOW_SERVICE_URL
+METAFLOW_DATASTORE_SYSROOT_S3
+WAGGLE_DANCE_DEFAULT_ENVIRONMENT
+WAGGLE_DANCE_DEV_ENDPOINT
+WAGGLE_DANCE_STAGE_ENDPOINT
+WAGGLE_DANCE_PROD_READ_ONLY_ENDPOINT
+WAGGLE_DANCE_PROD_ENDPOINT
+AWS_ROLE_ARN
+AWS_WEB_IDENTITY_TOKEN_FILE
+"""
+
+
+def identity_formatter(path: str, flow_parameters_json: dict) -> str:
+    return path
+
+
+"""
+@s3_sensor is implemented as a flow decorator that's used by customers to ensure 
+a workflow only begins after a key in S3 has been written to. 
+
+Example usage:
+
+    def formatter(path: str, flow_parameters_json: dict) -> str:
+        return path.replace("FLOW_ID", flow_parameters_json["FLOW_ID"])
+
+    @s3_sensor(
+        path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/$date/61/FLOW_ID",
```

Comment:
where is `$date` gotten from? Is this a variable available to the user?

Reply:
Sorry about that, we don't support `$date` anymore. Forgot to change this back.

Reply:
```python
# where you write the example of scheduled_time being a Flow parameter and would be within flow_parameters_json
# also document how all fields of flow_parameters_json are accessible
@s3_sensor(
        path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/date={scheduled_time}"

# how do I render s3://.../date=12-04-2021
@s3_sensor(
        path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/date={Date(scheduled_time).format('MM-DD-YY}')}")

```

```python
def foo(path: str, flow_parameters_json: dict) -> str:
    run_date = "12/2/2021"
    return path.format(run_date=run_date)

@s3_sensor(
        path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/date={run_date}"
        path_formatter=foo
```

```python
>>> path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/date={a}"
>>> path.format(a='hello')
's3://aip-example-sandbox/metaflow/S3SensorFlow/data/date=hello'
```

example:
```python
@s3_sensor(
        path="s3://aip-example-sandbox/metaflow/S3SensorFlow/data/year={Date(scheduled_time).year}/month={}/day={}"
```